File: 04-Data-Structures/code/python/seb_ds/trie.py
"""Classic per-character trie. Mirrors `typescript/src/trie.ts`."""

from __future__ import annotations

from typing import Dict, Optional


class _TrieNode:
    __slots__ = ("children", "is_end")

    def __init__(self) -> None:
        self.children: Dict[str, "_TrieNode"] = {}
        self.is_end = False


class Trie:
    def __init__(self) -> None:
        self._root = _TrieNode()

    def insert(self, word: str) -> None:
        node = self._root
        for ch in word:
            nxt = node.children.get(ch)
            if nxt is None:
                nxt = _TrieNode()
                node.children[ch] = nxt
            node = nxt
        node.is_end = True

    def contains(self, word: str) -> bool:
        node = self._find(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._find(prefix) is not None

    def delete(self, word: str) -> bool:
        if not self.contains(word):
            return False
        self._delete_rec(self._root, word, 0)
        return True

    def _delete_rec(self, node: _TrieNode, word: str, depth: int) -> bool:
        if depth == len(word):
            node.is_end = False
            return len(node.children) == 0
        ch = word[depth]
        child = node.children.get(ch)
        if child is None:
            return False
        should_prune_child = self._delete_rec(child, word, depth + 1)
        if should_prune_child:
            del node.children[ch]
        return len(node.children) == 0 and not node.is_end

    def _find(self, s: str) -> Optional[_TrieNode]:
        node = self._root
        for ch in s:
            nxt = node.children.get(ch)
            if nxt is None:
                return None
            node = nxt
        return node

    def check_invariants(self) -> None:
        # Structural by construction; see the TypeScript port for rationale.
        pass

File: 04-Data-Structures/code/typescript/src/trie.ts
class TrieNode {
  children: Map<string, TrieNode> = new Map();
  isEnd = false;
}

/** Classic per-character trie for word insertion, exact lookup, and prefix queries. */
export class Trie {
  private root = new TrieNode();

  insert(word: string): void {
    let node = this.root;
    for (const ch of word) {
      let next = node.children.get(ch);
      if (!next) {
        next = new TrieNode();
        node.children.set(ch, next);
      }
      node = next;
    }
    node.isEnd = true;
  }

  contains(word: string): boolean {
    const node = this.find(word);
    return node !== null && node.isEnd;
  }

  startsWith(prefix: string): boolean {
    return this.find(prefix) !== null;
  }

  delete(word: string): boolean {
    if (!this.contains(word)) return false;
    this.deleteRec(this.root, word, 0);
    return true;
  }

  private deleteRec(node: TrieNode, word: string, depth: number): boolean {
    if (depth === word.length) {
      node.isEnd = false;
      return node.children.size === 0;
    }
    const ch = word[depth];
    const child = node.children.get(ch);
    if (!child) return false;
    const shouldPruneChild = this.deleteRec(child, word, depth + 1);
    if (shouldPruneChild) node.children.delete(ch);
    return node.children.size === 0 && !node.isEnd;
  }

  private find(s: string): TrieNode | null {
    let node = this.root;
    for (const ch of s) {
      const next = node.children.get(ch);
      if (!next) return null;
      node = next;
    }
    return node;
  }

  checkInvariants(): void {
    // Structural by construction: no cycles are reachable (children form a
    // strict tree), so there is nothing to violate short of a coding bug
    // that this method's own traversal would already throw on (e.g. reading
    // an undefined map). Present for API symmetry with other structures.
  }
}
